causal-analysis/TowardsCausalThinking.ipynb

```diff
@@ -0,0 +1,2117 @@
+    "\n",
+    "In this document we explore the rationale of why we use the metrics we use, how are they related to each other and how the `knowledge of the way ACM works` is deeply related to all of these. We explore this through a series of equations.\n",
+    "\n",
+    "This document is split into 3 sections\n",
+    "\n",
+    "1. Capacity Planning - Green Field, Ideal Scenario\n",
+    "1. Capacity Planning - Green Field, Practical Scenario\n",
+    "1. Performance Analysis - on a running ACM \n",
+    "\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "be859f5a",
+   "metadata": {},
+   "source": [
+    "## Capacity Planning - Green Field, Ideal Scenario\n",
+    "\n",
+    "We list below a series of equations which needs to be solved to arrive at correct sizing decisions. The left hand side of these equations are the `dependendent variables` - aka what we want to measure or derive. And the right hand side variables drive to provide the answer. These are called `independent variables`.\n",
+    "\n",
+    "We call this ideal because you will notice that some of these `independent variables`below cannot be measured or is very hard to quantify. You will also notice that unless we make some approximations, the number of observations needed to do a robust calculation will be unbounded - hence not practical. We will be explaining these more in details under the following section `Capacity Planning - Practical Scenario`.\n",
+    "\n",
+    "Logically these `dependent variables` should be causally related to the `independent variables` - but for the purposes of practicality, we may have to make some simplifications/assumptions etc.\n",
+    "\n",
+    "<!--\n",
+    "### Observability\n",
+    "\n",
+    "- $Obs-size-for-thanos = \\mathbf{f}(number-of-managed-clusters, time-series-per-cluster^1)$\n",
+    "- $Obs-size-for-mco = \\mathbf{f}(number-of-managed-clusters)$\n",
+    "\n",
+    "### GRC\n",
+    "\n",
+    "- $Policy-size-for-controllers = \\mathbf{f}(number-of-managed-clusters, policy-count-per-cluster^1,policy-complexity^2,frequency-of-status-changes)$\n",
+    "\n",
+    "### Kube API /Etcd Server\n",
+    "\n",
+    "- $KubeAPI-size^3 = \\mathbf{f}(total-resources-created, frequency-of-resource-change, number-of-get-requests)$\n",
+    "- $total-resources-created = \\mathbf{f}(number-of-managed-clusters, acm-features-enabled, acm-release)^4$\n",
+    "- $number-of-get-requests = \\mathbf{f}(number-of-managed-clusters, acm-features-enabled, acm-release)^4$\n",
+    "- $Etcd-size = \\mathbf{f}(total-resources-created, frequency-of-resource-change)$\n",
+    "\n",
+    "Notes:\n",
+    "\n",
+    "1. For sake of Green field Capacity planning we do assume that policies or time series is uniform across all managed clusters.\n",
+    "1. Policy complexity is hard to express numerically\n",
```

Comment:
A lot of the complexity is basically the number of resources that need to be considered (due to ranges or namespaces searched) and the number of resources actually managed by the policy.